wallet: Change wallet validation order

w0xlt opened the pull request:
In the current code, the database is created before the last validation, which checks that passphrase is set and private keys are disabled.

Therefore, if this validation fails, it will result in an empty database and the user will not be able to recreate a wallet with the same name and with the correct parameters.

Behavior on the master branch:
```
$ ./src/bitcoin-cli -regtest -named createwallet wallet_name="invalid_wallet_01" disable_private_keys=true passphrase="passphrase"
error code: -4
error message:
Passphrase provided but private keys are disabled. A passphrase is only used to encrypt private keys, so cannot be used for wallets with private keys disabled.

$ ./src/bitcoin-cli -regtest -named createwallet wallet_name="invalid_wallet_01"
error code: -4
error message:
Wallet file verification failed. Failed to create database path '/home/w/.bitcoin/regtest/wallets/invalid_wallet'. Database already exists.
```

Behavior on the PR branch:
```
$ ./src/bitcoin-cli -regtest -named createwallet wallet_name="invalid_wallet_02" disable_private_keys=true passphrase="passphrase"
error code: -4
error message:
Passphrase provided but private keys are disabled. A passphrase is only used to encrypt private keys, so cannot be used for wallets with private keys disabled.

$ ./src/bitcoin-cli -regtest -named createwallet wallet_name="invalid_wallet_02"
{
  "name": "invalid_wallet_01",
  "warning": ""
}
```

w0xlt committed:
Change wallet validation order

In the current code, the database is created before the last validation,
which checks that passphrase is set and private keys are disabled.

Therefore, if this validation fails, it will result in an empty database
and the user will not be able to recreate a wallet with the same name
and with the correct parameters.

MarcoFalke commented:
It would be good to add a test for this if this is a bug fix

vincenzopalazzo reviewed (commented):
Concept ACK, the change looks trivial but I think to add a test unit will avoid trouble in the future if any change will introduce

w0xlt committed:
test: Add a test that creates a wallet with invalid parameters

Invalid parameters must not prevent a new wallet with the same name
from being created with the correct parameters

vincenzopalazzo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/pull/24859/commits/5db4e4fd7f4f2eab5b45cd0c4192ebb8f9328768

w0xlt commented:
Added https://github.com/bitcoin/bitcoin/pull/24859/commits/5db4e4fd7f4f2eab5b45cd0c4192ebb8f9328768 containing the test.

This test reproduces the behavior mentioned in the PR description and fails in the master branch.

aureleoules reviewed (commented):
tACK 5db4e4fd7f4f2eab5b45cd0c4192ebb8f9328768
Verified the behavior on master and this branch, nice fix!

MarcoFalke commented on test/functional/wallet_createwallet.py:
```diff
@@ -26,6 +26,12 @@ def run_test(self):
 
+        self.log.info("Run createwallet with invalid parameters.")
+        # Run createwallet with invalid parameters. This must not prevent a new wallet with the same name from being created with the correct parameters.
+        # See PR #24859 for details.
```
I don't think we need to add obscure `#nnn` to the source code. If anything this should be a full url, but I don't think it adds any value, so best to remove it.

vincenzopalazzo replied:
+1

w0xlt replied:
Done.

achow101 commented:
ACK 6f29409ad180ef00998ac05997f0fa03f98cd066
